receiver/httpdreceiver/scraper.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	rms := pdata.NewResourceMetricsSlice()
+	ilm := rms.AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
+	ilm.InstrumentationLibrary().SetName("otel/httpd")
+	now := pdata.NewTimestampFromTime(time.Now())
+
+	uptime := initMetric(ilm.Metrics(), metadata.M.HttpdUptime).Sum().DataPoints()
+	connections := initMetric(ilm.Metrics(), metadata.M.HttpdCurrentConnections).Gauge().DataPoints()
+	workers := initMetric(ilm.Metrics(), metadata.M.HttpdWorkers).Gauge().DataPoints()
+	requests := initMetric(ilm.Metrics(), metadata.M.HttpdRequests).Sum().DataPoints()
+	traffic := initMetric(ilm.Metrics(), metadata.M.HttpdTraffic).Sum().DataPoints()
+	scoreboard := initMetric(ilm.Metrics(), metadata.M.HttpdScoreboard).Gauge().DataPoints()
+
+	u, err := url.Parse(r.cfg.Endpoint)
+	if err != nil {
+		r.logger.Error("Failed to find parse server name", zap.Error(err))
+		return pdata.ResourceMetricsSlice{}, err
+	}
+	serverName := u.Hostname()
+
+	for metricKey, metricValue := range parseStats(stats) {
+		labels := pdata.NewAttributeMap()
+		labels.Insert(metadata.L.ServerName, pdata.NewAttributeValueString(serverName))
+
+		switch metricKey {
+		case "ServerUptimeSeconds":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				addToIntMetric(uptime, labels, i, now)
+			}
+		case "ConnsTotal":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				addToIntMetric(connections, labels, i, now)
+			}
+		case "BusyWorkers":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				labels.Insert(metadata.L.WorkersState, pdata.NewAttributeValueString("busy"))
+				addToIntMetric(workers, labels, i, now)
+			}
+		case "IdleWorkers":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				labels.Insert(metadata.L.WorkersState, pdata.NewAttributeValueString("idle"))
+				addToIntMetric(workers, labels, i, now)
+			}
+		case "Total Accesses":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				addToIntMetric(requests, labels, i, now)
+			}
+		case "Total kBytes":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				bytes := kbytesToBytes(i)
+				addToIntMetric(traffic, labels, bytes, now)
+			}
+		case "Scoreboard":
+			scoreboardMap := parseScoreboard(metricValue)
+			for identifier, score := range scoreboardMap {
+				labels.Upsert(metadata.L.ScoreboardState, pdata.NewAttributeValueString(identifier))
+				addToIntMetric(scoreboard, labels, score, now)
+			}
+		}
+	}
+
+	return rms, nil
+}
+
+// GetStats collects metric stats by making a get request at an endpoint.
+func (r *httpdScraper) GetStats() (string, error) {
+	url := fmt.Sprintf("%s%s", r.cfg.Endpoint, "/server-status?auto")
+	resp, err := r.httpClient.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// parseStats converts a response body key:values into a map.
+func parseStats(resp string) map[string]string {
```

Comment:
Consider creating a struct representing the result of parsing, so that all the magic strings in L109-L143 can move into this function, and the int parsing can happen here as well. Something like

```go
type serverStatus {
  serverUptimeSeconds int64
  connsTotal int64
  // etc
}
```

Reply:
Also here I would prefer not to introduce an intermediate data model for such a simple parsing operation. There's also the nuance of whether or not these values should be recorded at all if there is a parsing failure or if the key is not present. With an intermediate struct, we still have to differentiate between 0 and nil, which just adds additional complexity. Admittedly, I don't think the lines you mentioned are particularly great code, but it's straightforward enough to be easily understood. Of course, let me know if you feel strongly about this.

Reply:
Also here I would prefer not to add an intermediate data representation. A new data struct would still leave the question of how to to associate the magic strings with the appropriate data fields in the struct. There's also the nuance to whether or not a value was captured at all (since zero values do not communicate this). Such a structure or set of structures is not difficult to write, but it seems an unnecessary abstraction for such a simple parsing operation. Admittedly, this section is not what I would consider great code, but it seems to me sufficient for the problem. Of course, let me know if you feel strongly about this.